Add microshift preset

praveenkumar opened the pull request:
How to test

- Get the microshift bundle from internal server
- `crc config set preset microshift`
- `crc setup -b <microshift_bundle>`
- `crc start -b <microshift_bundle>`

Try to deploy a workload and check if everything is working.

Related to: https://github.com/crc-org/snc/pull/592

praveenkumar commented:
/hold

cfergeau commented on pkg/crc/constants/constants.go:
```diff
@@ -196,7 +198,7 @@ func GetDefaultCPUs(preset crcpreset.Preset) int {
 	case crcpreset.OpenShift, crcpreset.OKD:
 		return 4
-	case crcpreset.Podman:
+	case crcpreset.Podman, crcpreset.Microshift:
```
This makes me wonder if we can add these defaults to the bundle metadata instead of hardcoding this in crc

praveenkumar replied:
@cfergeau add default about preset and cpu/memory uses?

cfergeau replied:
Yes, each bundle could specify its expectations wrt to CPU/memory, rather than having an hardcoded list in crc. Some food for thought for future improvements.

cfergeau commented on pkg/crc/machine/bundle/metadata.go:
```diff
@@ -182,6 +182,8 @@ func (bundle *CrcBundleInfo) GetBundleType() crcPreset.Preset {
 	case "podman", "podman_custom":
 		return crcPreset.Podman
+	case "microshift", "microshift_custom":
+		return crcPreset.Microshift
```
Might be nicer as `type_str := strings.TrimSuffix(bundle.Type, "_custom")` and then the switch/case becomes
```
switch type_str {
  case "snc":
  case "podman:
 ...
 ```

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -369,8 +369,10 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
+	if !vm.bundle.IsMicroshift() {
+		if err := growRootFileSystem(sshRunner); err != nil {
+			return nil, errors.Wrap(err, "Error updating filesystem size")
+		}
```
Hopefully this is only a temporary limitation ?

praveenkumar replied:
@cfergeau yes, it should be temporary we need to find a way to extend the lvm with resized pv. I was able to do something following but that need `growpart` package to be installed and we might need a different helper for microshift preset for handing the resize.

```
   16  growpart /dev/vda 3
   17  lsblk 
   18  df -h
   19  pvs
   20  pvresize /dev/vda3
   21  pvs
   22  lvextend -r -l +100%FREE /dev/mapper/rhel-root 
```

cfergeau replied:
Maybe add a `// growRootFileSystem does not support LVM which is used by microshift images`

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -476,6 +479,31 @@ func (client *client) Start(ctx context.Context, startConfig types.StartConfig)
 		logging.Warn(fmt.Sprintf("Failed to query DNS from host: %v", err))
 	}
 
+	if vm.bundle.IsMicroshift() {
```
Fwiw, the more preset checks such as this one, the messier our code becomes. Especially in the already far too big `Start` method.
One possibel alternative is to try to add interfaces (here, could be a `Starter` interface, with a `PostVMStart` method maybe?), and have a `PodmanStarter`, `OpenShiftStarter`, ... which would all implement this interface.
This is a bit artificial here, especially as I haven't spent much time thinking about it, but it might apply better in other parts of the code.

gbraad commented:
@crc-org/crc-team  This has priority!

cfergeau commented on pkg/crc/machine/start.go:
```diff
@@ -915,3 +940,35 @@ func addPodmanSystemConnections(c *types.ConnectionDetails) error {
+		return err
+	}
+	if _, _, err := sshRunner.RunPrivileged("Change file permission of /opt/kubeconfig", "chmod", "0644", "/opt/kubeconfig"); err != nil {
+		return err
```
I'd use `CopyFile` since you just copied the file locally. It will do both copy to the right place and permission adjustments in one go.

cfergeau replied:
I don't understand the comment. `sshRunner.CopyFile(constants.KubeconfigFilePath, "/opt/kubeconfig", 0644)` does not work as expected? The first argument to `CopyFile` is a local file, not a file in the VM, so I don't understand why you mention the `core` user.

cfergeau commented on Makefile:
```diff
@@ -5,6 +5,7 @@ SHELL := /bin/bash
 OPENSHIFT_VERSION ?= 4.12.1
 PODMAN_VERSION ?= 4.3.1
 OKD_VERSION ?= 4.11.0-0.okd-2022-11-05-030711
+MICROSHIFT_VERSION ?= 4.12.1-202301261525.p0.g3db9e81.assembly.4.12.1.el8
```
Could this be just `4.12.1`? Or do we expect to need multiple bundle `subversions` for the same `4.12.1` version?

praveenkumar replied:
@cfergeau it would be `4.12.5`  not the full NVR of the rpm.

cfergeau commented on cmd/crc/cmd/start.go:
```diff
@@ -320,6 +339,23 @@ func writePodmanTemplatedMessage(writer io.Writer, s *startResult) error {
+		KubeConfigPath:    constants.KubeconfigFilePath,
+	})
+}
+
```
These cmd/start.go changes could be split to their own small commit at the end of the PR

praveenkumar replied:
In cmd/start.go only changes are related to microshift preset to generate the template, I am not sure if we split it into different commits because without `microshift` it doesn't have any usecase.

cfergeau replied:
This helps keeping each commit small and targeted, this one is mostly cosmetic, so this can be the last commit in the series.

cfergeau commented on pkg/crc/config/validations.go:
```diff
@@ -123,7 +123,7 @@ func ValidateYesNo(value interface{}) (bool, string) {
 	_, err := crcpreset.ParsePresetE(cast.ToString(value))
 	if err != nil {
-		return false, fmt.Sprintf("Unknown preset. Only %s and %s are valid.", crcpreset.Podman, crcpreset.OpenShift)
+		return false, fmt.Sprintf("Unknown preset. Only %s, %s and %s are valid.", crcpreset.Podman, crcpreset.OpenShift, crcpreset.Microshift)
```
I think I'd add a `preset: Add Microshift preset value` commit doing all the `case Microshift`, `if preset.OpenShift`, ... changes to separate these almost mechanical changes, but happening in multiple files , from the most important part of the PR, which is the changes in `pkg/machine/start.go`

You can still error out in `pkg/crc/config/validations.go` if trying to use the `microshift` preset so that it's not possible to try to use this preset before the commit with the `start.go` changes.

praveenkumar replied:
Just to have different commit without any functionality doesn't make sense if I put a different commit for preset specific changes and then error our as part of validation then again it is  dead code for that commit.

cfergeau replied:
Yes, these are a lot of small/trivial changes spread over a lot of files. If you move out to their own commit, this commit is quick to review, and the 'main commit' is smaller/more focused, which means easier to review.

praveenkumar commented:
https://github.com/crc-org/snc/pull/592 this is the dependencies for this PR because we need to create bundle for it.

praveenkumar commented:
/unhold

praveenkumar committed:
SSH: add helper to copy a file from instance

This is going to be used for copying kubeconfig file from
VM to host as part of microshift preset implementation.

praveenkumar committed:
Preset: Add microshift preset value

In next commit, this value is used for starting microshift
bundles as part of start.

praveenkumar committed:
Add microshift preset support

After this patch user can use `crc config set preset microshift` and
provision the cluster. For microshift preset we need to update the server
detail to kubeconfig because default it uses `127.0.0.1` when microshift
service start.

praveenkumar committed:
cmd/start: Add template for microshift preset

cfergeau commented on pkg/crc/machine/bundle/constants_test.go:
```diff
@@ -19,8 +19,10 @@ func TestURIOpenShiftVersion(t *testing.T) {
+			for p, remoteFile := range presetInfo {
+				if p == preset.OpenShift {
+					assert.True(t, strings.HasPrefix(remoteFile.URI, uriPrefix), fmt.Sprintf("URI %s does not match OpenShift version %s", remoteFile.URI, openshiftVersion))
+				}
```
This needs to be tested for microshift as well.

cfergeau commented on pkg/crc/machine/console.go:
```diff
@@ -18,7 +18,7 @@ func (client *client) GetConsoleURL() (*types.ConsoleResult, error) {
 	defer vm.Close()
 
-	if !vm.bundle.IsOpenShift() {
+	if vm.bundle.IsPodman() {
```
This could be moved to the 'preset' commit.

cfergeau commented on pkg/crc/preset/preset.go:
```diff
@@ -37,6 +37,8 @@ func ParsePresetE(input string) (Preset, error) {
 	case OKD.String():
 		return OKD, nil
+	case Microshift.String():
+		return Microshift, nil
```
This could go in the 'preset' commit.

praveenkumar replied:
This is preventing the earlier commit if user use preset microshift so that commit can error out. I think for this change this is much better location.

cfergeau replied:
Ah right, so that `validatePreset` errors out, I thought it could be done directly there ;)

cfergeau commented on pkg/crc/machine/kubeconfig.go:
```diff
@@ -227,3 +227,16 @@ func contains(arr []string, str string) bool {
 	return false
 }
+
+func updateServerDetailsToKubeConfig(input, output string) error {
```
I'd keep this one in the commit adding microshift support.

praveenkumar replied:
This is helper function and used for microshift preset type but it doesn't depend on the preset so created a separate commit.

cfergeau replied:
But having `updateServerDetails..` on its own is not really useful, it's an integral part of the microshift commit, and it's a single function in a single file, so it's not making the microshift commit a lot more complicated, which is why I'd keep it there. I can point a few small things which could be moved out of the microshift commit though if you want to make it a bit smaller.

cfergeau commented on pkg/crc/machine/bundle/metadata.go:
```diff
@@ -191,6 +193,14 @@ func (bundle *CrcBundleInfo) IsOpenShift() bool {
+
+func (bundle *CrcBundleInfo) IsPodman() bool {
+	return bundle.GetBundleType() == crcPreset.Podman
+}
```
We need to move *_away_* from these, not to add more such helpers ;)

cfergeau commented on pkg/crc/services/services.go:
```diff
@@ -12,4 +13,5 @@ type ServicePostStartConfig struct {
 	BundleMetadata bundle.CrcBundleInfo
 	IP             string
 	NetworkMode    network.Mode
+	Preset         preset.Preset
```
I don't think this is used.

praveenkumar replied:
I will remove it.

cfergeau commented on cmd/crc/cmd/start.go:
```diff
@@ -262,18 +262,37 @@ If you find an issue, please report it at https://github.com/openshift/okd
+	if runtime.GOOS == "windows" {
+		return fmt.Sprintf(tmpl, "$ENV:KUBECONFIG = \"{{ .KubeConfigPath }}\"")
+	}
+	return fmt.Sprintf(tmpl, "export KUBECONFIG={{ .KubeConfigPath }}")
```
I think this can use `shell.GetEnvString()`?

praveenkumar commented:
> I don't understand the comment. sshRunner.CopyFile(constants.KubeconfigFilePath, "/opt/kubeconfig", 0644) does not work as expected? The first argument to CopyFile is a local file, not a file in the VM, so I don't understand why you mention the core user.

@cfergeau the file is not local, since we are running the microshift service in the VM so it create the initial kubeconfig in the VM itself and that have `0600` permission for root user. so we need to put it to `/opt/kubeconfig` so the `ocConfig` can use it and also copy it from the VM to host.

cfergeau commented:
> > I don't understand the comment. sshRunner.CopyFile(constants.KubeconfigFilePath, "/opt/kubeconfig", 0644) does not work as expected? The first argument to CopyFile is a local file, not a file in the VM, so I don't understand why you mention the core user.
> 
> @cfergeau the file is not local, since we are running the microshift service in the VM so it create the initial kubeconfig in the VM itself and that have `0600` permission for root user. so we need to put it to `/opt/kubeconfig` so the `ocConfig` can use it and also copy it from the VM to host.

The line just before this code is `sshRunner.CopyFileFromVM("/var/lib/microshift/resources/kubeadmin/kubeconfig", constants.KubeconfigFilePath, 0600);` so we do have a local copy of this file at `constanst.KubeconfigFilePath`, don't we?

praveenkumar commented:
> > > I don't understand the comment. sshRunner.CopyFile(constants.KubeconfigFilePath, "/opt/kubeconfig", 0644) does not work as expected? The first argument to CopyFile is a local file, not a file in the VM, so I don't understand why you mention the core user.
> > 
> > 
> > @cfergeau the file is not local, since we are running the microshift service in the VM so it create the initial kubeconfig in the VM itself and that have `0600` permission for root user. so we need to put it to `/opt/kubeconfig` so the `ocConfig` can use it and also copy it from the VM to host.
> 
> The line just before this code is `sshRunner.CopyFileFromVM("/var/lib/microshift/resources/kubeadmin/kubeconfig", constants.KubeconfigFilePath, 0600);` so we do have a local copy of this file at `constanst.KubeconfigFilePath`, don't we?

@cfergeau ah spot on, I was looking only the VM side :(

cfergeau commented on pkg/crc/machine/bundle/constants_test.go:
```diff
@@ -19,8 +19,26 @@ func TestURIOpenShiftVersion(t *testing.T) {
+	}
+}
+
+func TestURIMicroshiftVersion(t *testing.T) {
```
Can't `TestURIOpenShiftVersion` be refactored to do both?

praveenkumar replied:
yes but better to have different for different url checks.

cfergeau replied:
If you want 2 different `TestURIXxxx`, fine with me, but please introduce a common helper for this, `testURIs(t, preset)` or such

cfergeau replied:
You could also introduce a [table-driven test](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests), but it's probably overkill here

cfergeau commented on pkg/crc/config/validations.go:
```diff
@@ -123,7 +123,7 @@ func ValidateYesNo(value interface{}) (bool, string) {
 	_, err := crcpreset.ParsePresetE(cast.ToString(value))
 	if err != nil {
-		return false, fmt.Sprintf("Unknown preset. Only %s and %s are valid.", crcpreset.Podman, crcpreset.OpenShift)
+		return false, fmt.Sprintf("Unknown preset. Only %s, %s and %s are valid.", crcpreset.Podman, crcpreset.OpenShift, crcpreset.Microshift)
```
You could add OKD to the string while you are at it.

cfergeau commented on cmd/crc/cmd/start.go:
```diff
@@ -262,18 +262,39 @@ If you find an issue, please report it at https://github.com/openshift/okd
 )
 
+func startTemplateForMicroshift() string {
+	tmpl := `Started the OpenShift edge cluster.
```
The product name is apparently "Red Hat build of MicroShift" https://access.redhat.com/documentation/en-us/red_hat_build_of_microshift/4.12 I'm not sure "OpenShift edge cluster" is an accurate description. Just `Started the MicroShift cluster`?

cfergeau replied:
The other name I've seen used is "Red Hat Device Edge" https://www.redhat.com/en/technologies/device-edge

praveenkumar replied:
@cfergeau this is something I need clarification from @gbraad because yesterday also we were discussing it, should it be `Openshift edge` or `microshift`. I am also bit align to use `microshift`.

gbraad replied:
We will use `microshift` as the preset name to choose it from our tooling, but what the UI does can be different. Perhaps we might add an alias to this later, but microshift is for now the choice.

As I understood it might be 'OpenShift for Edge', but this is not decided for our usecase.

cfergeau replied:
The question was about what to show in the UI, `microshift` as a preset name is not under discussion.
I'm not seeing any official use of "OpenShift for Edge" (and should it be "Red Hat OpenShift for Edge" then?)
I'm fine with `Started the MicroShift cluster` for now.

gbraad reviewed (approved):
preset: `microshift`

anjannath commented:
Tested on windows and macOS M1, in both the cases was able to start the `microshift` cluster and deploy simple workloads and interact with `oc`:

logs from M1:

```
❯ crc start -p ~/Downloads/pull-secret
INFO Checking if running as non-root
INFO Checking if crc-admin-helper executable is cached
INFO Checking if running on a supported CPU architecture
INFO Checking minimum RAM requirements
INFO Checking if crc executable symlink exists
INFO Checking if running emulated on a M1 CPU
INFO Checking if vfkit is installed
INFO Checking if old launchd config for tray and/or daemon exists
INFO Checking if crc daemon plist file is present and loaded
INFO Loading bundle: crc_microshift_vfkit_4.12.5_arm64...
INFO Creating CRC VM for Podman ...
INFO Generating new SSH key pair...
INFO CRC instance is running with IP 127.0.0.1
INFO CRC VM is running
INFO Updating authorized keys...
INFO Configuring shared directories
INFO Check internal and public DNS query...
INFO Check DNS query from host...
INFO Starting Microshift service... [takes around 1min]
INFO Waiting for kube-apiserver availability... [takes around 2min]
Started the OpenShift edge cluster.

export KUBECONFIG="/Users/anjan/.crc/machines/crc/kubeconfig"

Use the 'oc' command line interface:
  $ eval $(crc oc-env)
  $ oc COMMAND

❯ export KUBECONFIG="/Users/anjan/.crc/machines/crc/kubeconfig"

❯ eval `crc oc-env`

❯ oc get all -A
NAMESPACE                  NAME                                     READY   STATUS    RESTARTS      AGE
openshift-dns              pod/dns-default-hkmfv                    2/2     Running   0             25s
openshift-dns              pod/node-resolver-g9jr8                  1/1     Running   0             46s
openshift-ingress          pod/router-default-7c769bc888-vj5n2      1/1     Running   0             40s
openshift-ingress          pod/routes-controller                    1/1     Running   0             40s
openshift-ovn-kubernetes   pod/ovnkube-master-qwznh                 4/4     Running   0             45s
openshift-ovn-kubernetes   pod/ovnkube-node-jds5v                   1/1     Running   0             45s
openshift-service-ca       pod/service-ca-56587bb68d-588z9          1/1     Running   0             40s
openshift-storage          pod/topolvm-controller-9f7fdbcd6-qssh2   4/4     Running   0             46s
openshift-storage          pod/topolvm-node-dgkmq                   4/4     Running   1 (23s ago)   25s

NAMESPACE           NAME                              TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)                   AGE
default             service/kubernetes                ClusterIP   10.43.0.1     <none>        443/TCP                   72s
openshift-dns       service/dns-default               ClusterIP   10.43.0.10    <none>        53/UDP,53/TCP,9154/TCP    46s
openshift-ingress   service/router-internal-default   ClusterIP   10.43.11.67   <none>        80/TCP,443/TCP,1936/TCP   46s

NAMESPACE                  NAME                            DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
openshift-dns              daemonset.apps/dns-default      1         1         1       1            1           kubernetes.io/os=linux   46s
openshift-dns              daemonset.apps/node-resolver    1         1         1       1            1           kubernetes.io/os=linux   46s
openshift-ovn-kubernetes   daemonset.apps/ovnkube-master   1         1         1       1            1           kubernetes.io/os=linux   45s
openshift-ovn-kubernetes   daemonset.apps/ovnkube-node     1         1         1       1            1           kubernetes.io/os=linux   45s
openshift-storage          daemonset.apps/topolvm-node     1         1         1       1            1           <none>                   46s

NAMESPACE              NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
openshift-ingress      deployment.apps/router-default       1/1     1            0           46s
openshift-service-ca   deployment.apps/service-ca           1/1     1            1           46s
openshift-storage      deployment.apps/topolvm-controller   1/1     1            1           46s

NAMESPACE              NAME                                           DESIRED   CURRENT   READY   AGE
openshift-ingress      replicaset.apps/router-default-7c769bc888      1         1         1       46s
openshift-service-ca   replicaset.apps/service-ca-56587bb68d          1         1         1       46s
openshift-storage      replicaset.apps/topolvm-controller-9f7fdbcd6   1         1         1       46s

❯ oc create ns testcrc
namespace/testcrc created

❯ oc run testpod --image=nginx -n testcrc
Warning: would violate PodSecurity "restricted:v1.24": allowPrivilegeEscalation != false (container "testpod" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "testpod" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "testpod" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "testpod" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
pod/testpod created

❯ oc get pod -n testcrc
NAME      READY   STATUS    RESTARTS   AGE
testpod   1/1     Running   0          27s
```

gbraad commented:
/lgtm

cfergeau reviewed (changes requested):
https://github.com/crc-org/crc/pull/3421#discussion_r1139925995
